src/commonMain/kotlin/exchange.dydx.abacus/state/manager/notification/NotificationsProvider.kt

```diff
@@ -0,0 +1,74 @@
+package exchange.dydx.abacus.state.manager.notification
+
+import exchange.dydx.abacus.output.Asset
+import exchange.dydx.abacus.output.Notification
+import exchange.dydx.abacus.protocols.ParserProtocol
+import exchange.dydx.abacus.state.manager.V4Environment
+import exchange.dydx.abacus.state.manager.notification.providers.BlockRewardNotificationProvider
+import exchange.dydx.abacus.state.manager.notification.providers.OrderStatusChangesNotificationProvider
+import exchange.dydx.abacus.state.model.TradingStateMachine
+import exchange.dydx.abacus.utils.IMap
+import exchange.dydx.abacus.utils.JsonEncoder
+import exchange.dydx.abacus.utils.ParsingHelper
+import exchange.dydx.abacus.utils.UIImplementations
+import kollections.toIMap
+
+interface NotificationsProviderProtocol {
+    fun buildNotifications(
+        stateMachine: TradingStateMachine,
```

Comment:
I think using the full `TradingStateMachine` type will make it really annoying to test each notification provider in isolation.

Rather than passing it into the method, each provider should just take in the dependencies it needs in its constructor. Then, we can create lighter-weight interfaces for `TradingStateMachine` to implement that will allow us to easily make fakes to test against.

Reply:
The dependency is TradingStateMachine :-). We can pass it in from the constructor, but then the callers will have be updated which increases the scope of this PR.

Reply:
Passing it into the constructor allows the different implementations to more narrowly specify what they need from TradingStateMachine, rather than taking in the whole object which will make it easier to test. 

I don't think the signature updates are a significant scope increase - there are very few usages.

Reply:
The dependency is already passed from the constructor. 

NotificationProvider is designed to be a composite of all providers which have a common interface.  Each provider just operates on the interface, and you can write tests independently.  Without a common interface, tests will be more difficult to write since the logic is now split between the individual provider and composite provider.